File: examples/webhook_receiver.py
"""
AgentPay — Webhook Receiver Example

Receive real-time notifications when your agent spends, receives funds,
or hits spending limits. Webhooks use HMAC-SHA256 for verification.

Requirements:
    pip install fastapi uvicorn
"""
import hashlib
import hmac
import json
from fastapi import FastAPI, Request, HTTPException

app = FastAPI(title="AgentPay Webhook Receiver")

# Your webhook secret (set via POST /v1/webhook)
WEBHOOK_SECRET = "your-webhook-secret-here"


def verify_signature(payload: bytes, signature: str, secret: str) -> bool:
    """Verify HMAC-SHA256 webhook signature."""
    expected = hmac.new(
        secret.encode(), payload, hashlib.sha256
    ).hexdigest()
    return hmac.compare_digest(f"sha256={expected}", signature)


@app.post("/webhook/agentpay")
async def receive_webhook(request: Request):
    """Handle AgentPay webhook events."""
    body = await request.body()
    signature = request.headers.get("X-AgentPay-Signature", "")

    if not verify_signature(body, signature, WEBHOOK_SECRET):
        raise HTTPException(status_code=401, detail="Invalid signature")

    event = json.loads(body)
    event_type = event.get("event")
    data = event.get("data", {})

    match event_type:
        case "spend.completed":
            print(f"💸 Agent spent ${data['amount_usd']} — {data.get('description', 'N/A')}")

        case "spend.approval_needed":
            print(f"⚠️ Approval needed: ${data['amount_usd']} — {data.get('description')}")
            # Could forward to Slack, Discord, email, etc.

        case "deposit.completed":
            print(f"💰 Deposit received: ${data['amount_usd']}")

        case "balance.low":
            print(f"🔴 Low balance warning: ${data['balance_usd']} remaining")

        case _:
            print(f"📩 Unknown event: {event_type}")

    return {"received": True}


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=9000)

File: examples/webhook_receiver.ts
/**
 * AgentPay — TypeScript Webhook Receiver
 *
 * Express server that receives and verifies AgentPay webhooks
 * using the SDK's built-in HMAC-SHA256 signature verification.
 *
 * Requirements:
 *   npm install agentpay express
 */

import express from "express";
import { verifyWebhook } from "agentpay";

const WEBHOOK_SECRET = process.env.AGENTPAY_WEBHOOK_SECRET || "whsec_your_secret";
const PORT = 3000;

const app = express();

// IMPORTANT: use raw body for signature verification
app.use("/webhooks/agentpay", express.raw({ type: "application/json" }));

app.post("/webhooks/agentpay", (req, res) => {
  const signature = req.headers["x-agentpay-signature"] as string;
  const timestamp = req.headers["x-agentpay-timestamp"] as string;
  const body = req.body.toString();

  // Verify signature (timing-safe comparison, replay protection)
  if (!verifyWebhook(body, signature, WEBHOOK_SECRET, timestamp)) {
    console.warn("⚠️  Invalid webhook signature — rejecting");
    return res.status(401).json({ error: "Invalid signature" });
  }

  const event = JSON.parse(body);
  console.log(`✅ Verified event: ${event.type} (${event.id})`);

  // Handle event types
  switch (event.type) {
    case "transaction.completed":
      console.log(`💸 Spent $${event.data.amount} — ${event.data.description}`);
      break;
    case "transaction.refunded":
      console.log(`↩️  Refund $${event.data.amount_refunded}`);
      break;
    case "balance.low":
      console.log(`⚠️  Low balance: $${event.data.balance_usd}`);
      break;
    case "approval.requested":
      console.log(`🔐 Approval needed: $${event.data.amount} — ${event.data.description}`);
      break;
    default:
      console.log(`📨 Unhandled event type: ${event.type}`);
  }

  res.status(200).json({ received: true });
});

app.listen(PORT, () => {
  console.log(`🔔 Webhook receiver listening on port ${PORT}`);
});
